daemon/api_download.go

```diff
@@ -68,38 +108,154 @@ func postSnapDownload(c *Command, r *http.Request, user *auth.UserState) Respons
+		}
+		ss.resume = action.resumePosition
 	}
-	if len(sars) != 1 {
-		return InternalError("internal error: unexpected number %v of results for a single download", len(sars))
+
+	if !action.HeaderPeek {
+		r, s, err := theStore.DownloadStream(context.TODO(), action.SnapName, ss.Info, action.resumePosition, user)
+		if err != nil {
+			return InternalError(err.Error())
+		}
+		ss.stream = r
+		if s != 206 {
+			// store/cdn has no partial content (valid
+			// reply per RFC)
+			logger.Debugf("store refused our range request")
+			ss.resume = 0
+		}
 	}
-	info := sars[0].Info
 
-	downloadInfo := info.DownloadInfo
-	resume := int64(0)
-	r, status, err := getStore(c).DownloadStream(context.TODO(), action.SnapName, &downloadInfo, resume, user)
+	return ss
+}
+
+func newSnapStream(snapName string, info *snap.Info, secret []byte) (*snapStream, error) {
+	dlInfo := &info.DownloadInfo
+	fname := filepath.Base(info.MountFile())
+	tokenJSON := downloadTokenJSON{
+		SnapName: snapName,
+		Filename: fname,
+		Info:     dlInfo,
+	}
+	tokStr, err := sealDownloadToken(&tokenJSON, secret)
 	if err != nil {
-		return InternalError(err.Error())
+		return nil, err
 	}
-	// XXX: check for 206 as well here once we set resume to something other
-	// than "0"
-	if status != 200 {
-		return InternalError("internal error: unexpected status code from DownloadStream: %v", status)
+	return &snapStream{
+		SnapName: snapName,
+		Filename: fname,
+		Info:     dlInfo,
+		Token:    tokStr,
+	}, nil
+}
+
+func newResumingSnapStream(snapName string, tokStr string, secret []byte) (*snapStream, error) {
+	d, err := unsealDownloadToken(tokStr, secret)
+	if err != nil {
+		return nil, err
 	}
+	if d.SnapName != snapName {
+		return nil, fmt.Errorf("resume snap name does not match original snap name")
+	}
+	return &snapStream{
+		SnapName: snapName,
+		Filename: d.Filename,
+		Info:     d.Info,
+	}, nil
+}
+
+type downloadTokenJSON struct {
+	SnapName string             `json:"snap-name"`
+	Filename string             `json:"filename"`
+	Info     *snap.DownloadInfo `json:"dl-info"`
+}
+
+func sealDownloadToken(d *downloadTokenJSON, secret []byte) (string, error) {
+	b, err := json.Marshal(d)
+	if err != nil {
+		return "", err
+	}
+	mac := hmac.New(sha256.New, secret)
+	mac.Write(b)
+	tok := mac.Sum(b)
+	return base64.RawURLEncoding.EncodeToString(tok), nil
+}
 
-	return fileStream{
-		SnapName: action.SnapName,
-		Filename: info.Filename(),
-		Info:     downloadInfo,
-		stream:   r,
+var errInvalidDownloadToken = errors.New("download token is invalid")
+
+func unsealDownloadToken(tokStr string, secret []byte) (*downloadTokenJSON, error) {
+	tok, err := base64.RawURLEncoding.DecodeString(tokStr)
+	if err != nil {
+		return nil, errInvalidDownloadToken
+	}
+	sz := len(tok)
+	if sz < sha256.Size {
+		return nil, errInvalidDownloadToken
+	}
+	h := tok[sz-sha256.Size:]
+	b := tok[:sz-sha256.Size]
+	mac := hmac.New(sha256.New, secret)
+	mac.Write(b)
+	if !hmac.Equal(h, mac.Sum(nil)) {
```

Comment:
Documentation states that this is a timing side channel. I think we should explicitly acknowledge this and document why we think it is okay in our case.

Reply:
documentation says using the helper avoids the side channel (to an extent). We are trying to mostly avoid casual tampering here, also if one is root on the machine all of this is of course is moot